### PR TITLE
C++: Improve PrintAST performance.

### PR DESCRIPTION
**Pull request by criemen**
Faster PrintAST with less and easier to understand code!

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/Type.qll [hunk 1]
```diff
@@ -1710,28 +1710,6 @@ class AutoType extends TemplateParameter {
...
-        res = spec + " " + rest and
-        internalSpecString(t, rest, i + 1)
-      )
-    else (
-      allSpecifiers(i, _) and internalSpecString(t, res, i + 1)
-    )
-  )
-  or
-  i = count(Specifier s) + 1 and res = ""
-}
```
Comment (geoffw0) [inline cpp/ql/src/semmle/code/cpp/Type.qll:1733]: Was this just ... very old code?  (we've had `concat` for a while now!)

Comment (criemen) [reply cpp/ql/src/semmle/code/cpp/Type.qll:1733]: I didn't dig in the history - I suspect either it's very old or written by somebody who was fairly new to QL at the time.

**Comment by geoffw0** [comment]
How did you check performance?

**Comment by criemen** [comment]
Manually on a single file from facebookincubator/fizz, the previous code took about 20sec, the new code takes about 100ms.

File: cpp/ql/src/semmle/code/cpp/Type.qll [hunk 2]
```diff
@@ -1308,7 +1308,7 @@ class SpecifiedType extends DerivedType {
    * only depends on the specifiers, not on the source program). This is intended
    * for debugging queries only and is an expensive operation.
    */
-  string getSpecifierString() { internalSpecString(this, result, 1) }
+  string getSpecifierString() { result = concat(this.getASpecifier().getName(), " ") + " " }
```
Comment (geoffw0) [inline cpp/ql/src/semmle/code/cpp/Type.qll:1311]: How do you feel about removing the trailing space, and fixing up the three? places this predicate is used?

File: cpp/ql/src/semmle/code/cpp/Type.qll (hunk 1 above)
Comment (dbartol) [reply cpp/ql/src/semmle/code/cpp/Type.qll:1733]: Hrm. I assume that I wrote this, but it doesn't look like my style. In any case, nice improvement in both readability and performance.

**Review by dbartol — APPROVED** [review]
LGTM

File: cpp/ql/src/semmle/code/cpp/Type.qll (hunk 2 above)
Comment (criemen) [reply cpp/ql/src/semmle/code/cpp/Type.qll:1311]: We could do that, but in theory it is an API change.
It is rectified by the fact that the predicate is only for debugging, and that the whitespace behaviour is not documented, but still, at least in my opinion it's not a no-brainer.

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/Type.qll:1311]: I'm happy to accept this API change in a predicate that has "This is intended for debugging queries only" in its QLDoc.

This is a good reminder that we should be conscious about which APIs we make public. The QLDoc for `getSpecifierString` should really start with "INTERNAL: do not use.", and ideally that predicate should not reside in `Type.qll` at all.

**Pull request merged**